Initial OTAPI3 port / .NET6

SignatureBeef opened the pull request:
~~DO NOT MERGE YET. This is a draft, and SQLite is not yet working and hence why I would like to create a discussion around this.~~

This PR upgrades TShock to OTAPI3 and .NET5 (6 later on). This is a major/breaking upgrade.

- No testing has occurred at all and once merged into the otapi3 branch it should have full end-end testing. 
- It must not go to general-devel until the entire team is satisfied, and TSAPI would need to do this first as well.
- Packages have updated, need to ensure thy work e.g MySql (havent had a chance to set this up)
- AssemblyInfo.cs was replaced to the .csproj, as it is now shared between the assembly info and nuget package info from there.
- CI is completely broken. The new stack is much easier to build, but we also need to figure out the desired publish settings too (and what about cross platform etc).
- ~~**Currently Mono.Data.Sqlite is used and the sqlite3.dll needs updating to x64, or we need to switch to System.Data.Sqlite which i can confirm already works with a assembly resolver, or Microsoft.Data.Sqlite**~~


~~What preference does the team have for the sqlite library on net5 onward? This is my main blocking question at this point.~~

SignatureBeef committed:
Initial OTAPI3 port

TODO: sqlite

hakusaro commented:
> What preference does the team have for the sqlite library on net5 onward? This is my main blocking question at this point.

I would say`System.Data.Sqlite`.

SignatureBeef committed:
Implement System.Data.SQLite

TODO: determine if we can do something better than a resolver

SignatureBeef committed:
Correct nuget package to GPL-3.0-or-later

hakusaro commented on TShockAPI/TShockAPI.csproj:
```diff
@@ -1,242 +1,71 @@
+    <AssemblyVersion>4.5.7</AssemblyVersion>
+    <FileVersion>4.5.7</FileVersion>
+    <AssemblyTitle>TShock for Terraria</AssemblyTitle>
+    <Company>Re-Logic, Pryaxis &amp; TShock Contributors</Company>
```
```suggestion
    <Company>Pryaxis &amp; TShock Contributors</Company>
```

I think we can drop RL from this because we aren't packing their stuff here anymore. I think this copyright was vestigial from back in the "we are shipping their source code with TShock" days.

SignatureBeef replied:
that one is up to you, so sure? this was just moved from AssemblyInfo.cs

hakusaro replied:
@DeathCradle it's your branch :P

hakusaro commented on TShockAPI/TShockAPI.csproj:
```diff
@@ -1,242 +1,71 @@
+    <AssemblyTitle>TShock for Terraria</AssemblyTitle>
+    <Company>Re-Logic, Pryaxis &amp; TShock Contributors</Company>
+    <Product>TShockAPI</Product>
+    <Copyright>Copyright © Re-Logic, Pryaxis &amp; TShock Contributors 2011-2021</Copyright>
```
```suggestion
    <Copyright>Copyright © Pryaxis &amp; TShock Contributors 2011-2021</Copyright>
```

hakusaro commented on TShockAPI/TShockAPI.csproj:
```diff
@@ -1,242 +1,71 @@
-      <ProductName>Windows Installer 3.1</ProductName>
-      <Install>true</Install>
-    </BootstrapperPackage>
+    <Content Include="$(OutputPath)\TerrariaServerAPI.dll" IncludeInPackage="true" Pack="true" PackagePath="lib\$(TargetFramework)" />
```
Should I be asking why we're including TSAPI in the package?

SignatureBeef replied:
plugin refs/hooks, since tsapi has no nuget package. though, now is the time if you want to do that...

hakusaro replied:
@DeathCradle I mean, I think we should just start publishing these as nuget packages but I was unsure if we should move forward with merging this before that

SignatureBeef replied:
we can probably do it for TSAPI as the otapi3 branch is on the main repo now, so i can likely get it to produce the package and push to nuget once we have an account and tokens setup.
tshock would be after this merge yeh.
edit: only seen your latest comment, feel free to do CI, i certainly wont fight you for it

hakusaro replied:
Kk. Sounds like I should coordinate with you about making a Pryaxis org and namespace for nuget stuff.

hakusaro commented on TShockAPI/TShockAPI.csproj:
```diff
@@ -1,242 +1,71 @@
+    <Copy SourceFiles="$(OutputPath)\TShockAPI.deps.json" DestinationFolder="$(SolutionDir)\TerrariaServerAPI\TerrariaServerAPI\bin\$(ConfigurationName)\net5.0\ServerPlugins" SkipUnchangedFiles="false" />
+    <Copy SourceFiles="$(OutputPath)\TShockAPI.pdb" DestinationFolder="$(SolutionDir)\TerrariaServerAPI\TerrariaServerAPI\bin\$(ConfigurationName)\net5.0\ServerPlugins" SkipUnchangedFiles="false" />
+    <Copy SourceFiles="$(OutputPath)\TShockAPI.xml" DestinationFolder="$(SolutionDir)\TerrariaServerAPI\TerrariaServerAPI\bin\$(ConfigurationName)\net5.0\ServerPlugins" SkipUnchangedFiles="false" />
+    <Copy SourceFiles="$(OutputPath)\System.Data.SQLite.dll" DestinationFolder="$(SolutionDir)\TerrariaServerAPI\TerrariaServerAPI\bin\$(ConfigurationName)\net5.0\ServerPlugins" SkipUnchangedFiles="false" />
```
Is there any way to keep non-plugins from the `ServerPlugins` folder? One of my biggest pet peeves with netcore/dotnet 5 is that the build outputs are extremely messy. For the purposes of simplicity, I think I'd prefer if we could throw these dlls somewhere that aren't `ServerPlugins` or the root directory, but if we can't it's fine.

SignatureBeef replied:
potentially, would need to investigate. e.g. runtimes in debug mode contain the native deps, and we can control the resolution via code.

hakusaro reviewed (changes requested):
I have a couple minor suggestions and a couple questions, other than that this looks okay code-wise.

The only other thing that I kind of want to require is that CI works, because we do rely on CI for basically everything. The current CI outputs are directly used for releases and package everything in the exact correct way it needs to be to ship a release out. Specifically, you can just drag and drop everything from a release onto an existing server to upgrade, which means that every output is packaged correctly in terms of location in `ServerPlugins` or not, and vice versa.

The complication with .net5 is that we need to start shipping assemblies for macOS, Linux, and Windows I guess. Like I already commented, I think the biggest thing is that when we package these for release, we need to make sure that there isn't a billion DLLs in the release folder people download if at all possible. Can we just move everything into a `System` or something folder?

Honestly, we probably need outputs for:
* x64 win
* x64 macOS
* x64 linux
* arm64e macOS
* arm64 linux

I'm not sure if you want to address this toolchain stuff in this PR, or just merge it and then work on subsequent work to get other things done.

**I have not tested this locally yet, but I will do so, soon.**

SignatureBeef commented on TShockAPI/TShockAPI.csproj:
```diff
@@ -1,242 +1,71 @@
+    <Product>TShockAPI</Product>
+    <Copyright>Copyright © Re-Logic, Pryaxis &amp; TShock Contributors 2011-2021</Copyright>
+
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
```
this is why there is a bunch of assemblies. only currently needed for debugging automation (ie copying deps to tsapi so it actually runs).
however, im sure we can streamline this process, and we might need to work out our deployment strategy...which im unsure the best course of action yet. (@tylerjwatson is the legend on this stuff i would assume but he's hella busy)

hakusaro replied:
Fair

hakusaro commented:
I think I would articulate that:

1. In order for me to approve merging it, I need to be able to build locally and test on macOS 12.0.1 (21A559) (arm64e on Rosetta), and currently, I can't do that, and I'm not quite sure why. I just kind of yolo'd this build process, though. If you have better insight on how to build, please let me know.
2. We don't have to get everything right on the merge to `otapi3` given that `general-devel` is the main branch. However, in order to merge to general-devel, most of the issues that I outlined above need to be fixed/rectified. For example, I can work on CI/tooling, or you can, but I don't want to consider that a blocker for a branch on TShock. We can easily just merge the current branch and assemble CI/releases in bits and bobs until we're happy to merge to general-devel.

The timeline of events would look something like this:
1. Merge to otapi3.
2. Get CI working.
3. Publish pre-release builds for people to test with on test servers and test plugins against.
4. Merge to general-devel, so that mainline CI starts occurring.
5. Prepare documentation on what to do/how to do it for things.
6. Publish a freaking docker file or otherwise update Ryshe's docker file so that docker users aren't left out in the cold.
7. Update the readme.
8. Release the stuff.

Currently, 1 is where we're at but we could compress 2 into 1. It just depends on how much work you want to do.

hakusaro commented:
Optional steps:
1. Get cute anime girl mascot for tshock (tbd)
2. Figure out how to get plugin developers to update (tbd)

SignatureBeef commented:
> 1. In order for me to approve merging it, I need to be able to build locally and test on macOS 12.0.1 (21A559) (arm64e on Rosetta), and currently, I can't do that, and I'm not quite sure why. I just kind of yolo'd this build process, though. If you have better insight on how to build, please let me know.

agreed. i will do more testing on osx/linux as time permits. possibly i need to build some initial instructions as there's a few ways to build it (cli / vs / other editors etc), and as we found out earlier they don't all work the same way. i think i just need to spend some extra time on getting it setup for all platforms correctly.

> 2. We don't have to get everything right on the merge to `otapi3` given that `general-devel` is the main branch. However, in order to merge to general-devel, most of the issues that I outlined above need to be fixed/rectified. For example, I can work on CI/tooling, or you can, but I don't want to consider that a blocker for a branch on TShock. We can easily just merge the current branch and assemble CI/releases in bits and bobs until we're happy to merge to general-devel.

yeh honestly, i wont fight you to do the CI so go nuts.
i think tsapi needs the csproj settings to generate a package, and the settings moved from AssemblyInfo like i done for tshock here so it's shared. 
feel free to use some of the otapi ci action stuff as a reference, ie for building dotnet, and uploading (not building nuget as otapi is manually done as patchers are a pain).

SignatureBeef committed:
Improve native resolutions

SignatureBeef committed:
VS+Code launch profile improvements

hakusaro commented on .vscode/launch.json:
```diff
@@ -0,0 +1,24 @@
+            "request": "attach"
+        }
+    ]
+}
```
Missing EOF newline, can you please add?

hakusaro commented on .vscode/tasks.json:
```diff
@@ -0,0 +1,17 @@
+            "problemMatcher": "$msCompile"
+        }
+    ]
+}
```
Missing EOF newline, can you please add?

SignatureBeef replied:
Sure I guess, for my own education though, why? These files can be auto generated, so not sure if new lines can be accidentally lost. I don't see us regenerating them for a while though, if anything I'd add to it.

hakusaro replied:
@DeathCradle because if they don't have newlines they don't meet the POSIX definition of a line, which will break random unix utilities in unexpected ways (e.g., they will read to the 2nd to last line and stop): https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206

SignatureBeef replied:
@hakusaro these are vscode files specifically which is why i asked, and it handles them fine on all platforms ive tested so far. (im aware of the unix stuff for other things though)
nonetheless, ill add them if i don't forget

hakusaro commented on TShockAPI/TShock.cs:
```diff
@@ -190,25 +191,74 @@ public TShock(Main game)
+			}
+			else if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+			{
+				var x64 = Path.Combine(Environment.CurrentDirectory, "runtimes", "win-x64");
```
I'm guessing this resolver needs to be updated for arm64?

SignatureBeef replied:
potentially, can't say i have such a device to test with yet though. but might be able to figure something out.

hakusaro replied:
At least with arm64, the biggest thing I'm thinking about is raspberry pi users who are using mono on rpi. I'm not sure if you have one kicking around?

SignatureBeef replied:
uh actually i do! rpi 4 model b is the best i have though, is that relevant? havent looked at these for years.

hakusaro replied:
Yeah that works, should be valid.

SignatureBeef replied:
update: ive tested my rpi, and OTAPI itself runs, however ive discovered MonoMod issues where the runtime events don't -always- work, potentially JIT issues wrt runtime detours.

hakusaro reviewed (changes requested):
I think this is for the most part the same as my previous review, but with added missing EOFs. I haven't yet tested this locally.

hakusaro commented:
Quoting Death from Discord:

```bash
dotnet build TShock.sln
cd TShockAPI/bin/Debug/net5.0
dotnet ./TerrariaServerAPI.dll
```

hakusaro commented:
Okay, to be clear, this does at least boot on my machine, but I don't think that `dotnet ./TerrariaServerAPI.dll` is very friendly to...most people. I guess maybe it is? Is it?

SignatureBeef commented:
> Okay, to be clear, this does at least boot on my machine, but I don't think that `dotnet ./TerrariaServerAPI.dll` is very friendly to...most people. I guess maybe it is? Is it?

For dev should be if you want to do it CLI style. Other than that use vs code or vs and it's 1 click. 
Release/published version is another story, but will figure that out after the above changes are done (I'll get back to these after the net6 update, which isn't far away now that I can debug the client errors making the net6 otapi release worthy)

SignatureBeef committed:
Merge branch 'general-devel' into otapi3

SignatureBeef committed:
Switch to .net6

SignatureBeef committed:
Update company/copyright + add new lines to .vscode files

SignatureBeef committed:
Update submodule for ./bin folder changes

This should move most dlls from the root to a ./bin folder.

hakusaro commented:
![Screen Shot 2021-11-29 at 22 23 36](https://user-images.githubusercontent.com/532647/143996466-36003a1e-fc08-4352-8617-2d861e59fb6e.png)

It's working on .NET 6 on macOS (this is under Rosetta)!

hakusaro reviewed (approved):
I'm gonna say lgtm because it runs + compiles on my machine in an easy way. Will continue to incrementally review as commits are added.

SignatureBeef committed:
Add TShock Launcher project (TShock.exe) with ./bin folder for dependencies

+ submodule update
TODO: retest on osx and see if we can include file permissions

SignatureBeef committed:
Improve OSX non self contained mode wrt name/folder clashing

SignatureBeef committed:
Correct windows launching in debug mode, and dont move the TShock.sh in osx

SignatureBeef committed:
Add ubuntu runtime identifier

hakusaro commented on TShockLauncher/TShock.sh:
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+./bin/TShock.Run
```
Can you add a newline to this file?

SignatureBeef replied:
done https://github.com/Pryaxis/TShock/pull/2509/commits/c159f2b388a3e084449ba089d0e11911517ec03e

hakusaro reviewed (changes requested):
Another missing newline :x

SignatureBeef committed:
Add newline to TShock.sh + add linux-arm64 to launcher release

SignatureBeef commented:
> Quoting Death from Discord:
> 
> ```shell
> dotnet build TShock.sln
> cd TShockAPI/bin/Debug/net5.0
> dotnet ./TerrariaServerAPI.dll
> ```

### update since launcher was added:
```shell
dotnet build TShock.sln
cd TShockLauncher/bin/Debug/net6.0
dotnet TShock.dll
```

### publishing:
win
```shell
cd TShockLauncher
dotnet publish -r win-x64 -f net6.0 -c Release -p:PublishSingleFile=true --self-contained true
cd bin/Release/net6.0/win-x64/publish
TShock.exe
```
osx
```shell
cd TShockLauncher
dotnet publish -r osx-x64 -f net6.0 -c Release -p:PublishSingleFile=true --self-contained true
cd bin/Release/net6.0/osx-x64/publish
./TShock.sh 
```
linux
```shell
cd TShockLauncher
dotnet publish -r linux-x64 -f net6.0 -c Release -p:PublishSingleFile=true --self-contained true
cd bin/Release/net6.0/linux-x64/publish
./TShock.sh 
```

### published/release folder structure:
![image](https://user-images.githubusercontent.com/776327/144367238-085494b0-6fa1-4793-9c99-37002633a0a3.png)
bin contains all dlls (OTAPI.dll, TerrarriaServer.dll - api exists here when publishsinglefile=false)
serverplugins remains the same, TShock.dll exists in there
TShock.[exe|sh] is the new launcher

SignatureBeef committed:
Single file publish support, test project and simple build CI

This also adds remote raspberry pi debugging with default install details. More testing is required as MonoMod may not be working for arm64 still

CI might not work yet either

SignatureBeef committed:
Add recursive checkout option to ci

SignatureBeef committed:
Correct ci matrix

Nothing to see here...

SignatureBeef committed:
Publish launcher only

SignatureBeef committed:
CI upload artifacts + set tshock version to beta

SignatureBeef committed:
Remove build.yml + try and implement chmod in otapi3 ci

SignatureBeef committed:
CI - tar release to preserve permissions

Hopefully...

hakusaro commented:
>  update: ive tested my rpi, and OTAPI itself runs, however ive discovered MonoMod issues where the runtime events don't -always- work, potentially JIT issues wrt runtime detours.

Is this still valid?

hakusaro commented on TShockLauncher/Program.cs:
```diff
@@ -0,0 +1,66 @@
+﻿/*
+TShock, a server mod for Terraria
+Copyright (C) 2011-2019 Pryaxis & TShock Contributors
```
```suggestion
Copyright (C) 2021 Pryaxis & TShock Contributors
```

hakusaro reviewed (approved):
There's a very minor nit in here: copyright on the launcher should technically be 2021-x because it didn't exist before 2021.

The only quibble I have with the current CI output is that people aren't going to expect to download a file that contains like 7 levels deep of stuff. That can be simplified later, though, in all honesty, for public releases.

Everything else virtually lgtm. I'm not sure if I should worry or care about assembly resolving stuff + comments. I think @QuiCM is a better judge on if we should document that more.

So, lgtm. Thank you for all of your hard work on this!

I reckon we can merge this to tshock as an `otapi3` branch and then target a full release sometime around the second half of December.

![lgtm_trick](https://user-images.githubusercontent.com/532647/144976854-9af1cf2d-9141-4d2f-9fa7-b51560a4f8a2.gif)

SignatureBeef commented:
> > update: ive tested my rpi, and OTAPI itself runs, however ive discovered MonoMod issues where the runtime events don't -always- work, potentially JIT issues wrt runtime detours.
> 
> Is this still valid?

yes actually :c i havent been able to get back and test/debug it further. i/we also need to try other OS types on the rpi too.

SignatureBeef committed:
Update TShockLauncher/Program.cs

Co-authored-by: Lucas Nicodemus <shank@shanked.me>
